docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,242 @@
+Horizontal Pod Autoscaler with Arbitary Metrics
+===============================================
+
+The current Horizontal Pod Autoscaler object only has support for CPU as
+a percentage of requested CPU.  While this is certainly a common case, one
+of the most frequently sought-after features for the HPA is the ability to
+scale on different metrics (be they custom metrics, memory, etc).
+
+The current HPA controller supports targeting "custom" metrics (metrics
+with a name prefixed with "custom/") via an annotation, but this is
+suboptimal for a number of reasons: it does not allow for arbitrary
+"non-custom" metrics (e.g. memory), it does not allow for metrics
+describing other objects (e.g. scaling based on metrics on services), and
+carries the various downsides of annotations (not be typed/validated,
+being hard for a user to hand-construct, etc).
+
+Object Design
+-------------
+
+### Requirements ###
+
+This proposal describes a new version of the Horizontal Pod Autoscaler
+object with the following requirements kept in mind:
+
+1. The HPA should continue to support scaling based on percentage of CPU
+   request
+
+2. The HPA should support scaling on arbitrary metrics associated with
+   pods
+
+3. The HPA should support scaling on arbitrary metrics associated with
+   other Kubernetes objects in the same namespace as the HPA (and the
+   namespace itself)
+
+4. The HPA should make scaling on multiple metrics in a single HPA
+   possible and explicit (splitting metrics across multiple HPAs leads to
+   the possibility of fighting between HPAs)
+
+### Specification ###
+
+```go
+type HorizontalPodAutoscalerSpec struct {
+    // the target scalable object to autoscale
```

Comment:
Please change all of the field/struct comments to be regular sentences (e.g. start with capital letter and end with `.`). It will be more consistent with all other types definition. I know it's not the case currently, but we can fix it at the same time :)